diff -Naubr qt4urpm-1.0.1.orig/qt4urpm.cpp qt4urpm-1.0.1/qt4urpm.cpp
--- qt4urpm-1.0.1.orig/qt4urpm.cpp	2014-04-01 14:19:56.272645899 +0400
+++ qt4urpm-1.0.1/qt4urpm.cpp	2014-04-01 14:17:00.000000000 +0400
@@ -20,6 +20,7 @@
 #include <QAction>
 #include <QDesktopServices>
 #include <QMouseEvent>
+#include <QTemporaryFile>
 
 using namespace std;
 
@@ -31,6 +32,8 @@
 	ui->setupUi(this);
 	ui->orph_results->setColumnWidth(0,505);
 	ui->find_results->setColumnWidth(0,505);
+    connect(ui->trs_update,       SIGNAL(clicked()), this, SLOT(list_transactions()));
+    connect(ui->trs_rollback,     SIGNAL(clicked()), this, SLOT(rollback()));
 	connect(ui->orph_search,      SIGNAL(clicked()), this, SLOT(orphansearch()));
 	connect(ui->orph_notOrphaned, SIGNAL(clicked()), this, SLOT(notOrphAct()));
 	connect(ui->orph_remove,      SIGNAL(clicked()), this, SLOT(removeAct()));
@@ -42,8 +45,11 @@
 	connect(&proc, SIGNAL(finished(int, QProcess::ExitStatus)), this, SLOT(procfinish()));
 	connect(ui->orph_results, SIGNAL(cellEntered(int,int)), this, SLOT(changeTooltip(int,int)));
 
-	// Let's fill the list
+    procout = tmpnam(NULL);
+
+    // Let's fill the lists
 	orphansearch();
+    list_transactions();
 }
 
 /// destructor
@@ -52,6 +58,8 @@
 	// Better safer than safe
 	if (ui)
 		delete ui;
+
+    QFile::remove(procout);
 }
 
 /**
@@ -66,7 +74,32 @@
 	termlog(command.toStdString());
 	ui->urpm_status->setText(tr("Collecting data ..."));
 	procmode = ORPHANSEARCH;
+    proc.setStandardOutputFile(procout);
+	proc.start(command);
+    proc.waitForFinished(-1);
+    processData();
+    procmode = NONE;
+}
+
+/**
+ * list_transactions()
+ * list transactions that can be reverted
+ */
+void qt4urpm::list_transactions()
+{
+    if (procmode != NONE) {
+        proc.close();
+    }
+//    QString command("find /var/spool/repackage -type d | sort -n | head -2 | tail -1 | while read a; do b=`basename $a`; ((b=$b-1)); date --date @$b \"+%Y-%m-%d %H:%M:%S\"; done | while read last_date; do urpmi.recover --list \"$last_date\"; done");
+    QString command("urpmi.recover --list-safe");
+    termlog(command.toStdString());
+    ui->urpm_status->setText(tr("Collecting data ..."));
+    procmode = LISTTRANSACTIONS;
+    proc.setStandardOutputFile(procout);
 	proc.start(command);
+    proc.waitForFinished(-1);
+    processData();
+    procmode = NONE;
 }
 
 /**
@@ -89,13 +122,98 @@
 	termlog(command.toStdString());
 	ui->urpm_status->setText(tr("Collecting data ..."));
 	procmode = PACKAGESEARCH;
+    proc.setStandardOutputFile(procout);
 	proc.start(command);
+    proc.waitForFinished(-1);
+    processData();
+    procmode = NONE;
 }
 
 /// processes the output of a proc.start()
 void qt4urpm::processData()
 {
+    QString trans_date;
+    QString trans_pkgs;
+    int i, row;
+
+    QFile file(procout);
+    if (!file.open(QIODevice::ReadOnly)) {
+        return;
+    }
+
+    QTextStream in(&file);
+
 	switch (procmode) {
+        case LISTTRANSACTIONS:
+            // Cleanup
+            while (transactions.size())
+                transactions.pop_back();
+
+            // Begin of processing the data
+            termlog("Processing ...");
+            ui->urpm_status->setText(tr("Processing data ..."));
+
+            row=0;
+            i=0;
+
+            while (!in.atEnd()) {
+                QString newtrans = in.readLine();
+                termlog(newtrans.toStdString());
+                // Skip first two lines
+                i++;
+                if (i <= 2) {
+                    continue;
+                }
+                                newtrans = newtrans.trimmed();
+                termlog(newtrans.toStdString());
+                QStringList trans_elems = newtrans.split(" ");
+
+                // Launch urpmi.recover --list-safe to get an idea of how its output looks like.
+                // Here we want to combine all packages correcponding to the same date in one cell.
+                if (trans_elems.length() > 1) {
+                    if (!trans_date.isEmpty()) {
+                        // Dump previously collected entries
+                        ui->recover_list_transactions->setRowCount(row+1);
+                        transactions.push_back(trans_date);
+                        ui->recover_list_transactions->setItem(row, 0, new QTableWidgetItem(trans_date));
+                        ui->recover_list_transactions->setItem(row, 1, new QTableWidgetItem(trans_pkgs));
+                        // Do not allow the user to change the line - just selection.
+                        ui->recover_list_transactions->item(row, 0)->setFlags
+                            (Qt::ItemIsEnabled | Qt::ItemIsSelectable);
+                        ui->recover_list_transactions->item(row, 1)->setFlags
+                            (Qt::ItemIsEnabled | Qt::ItemIsSelectable);
+                        row++;
+                    }
+
+                    trans_date = trans_elems[0] + " " + trans_elems[1];
+                    trans_pkgs = trans_elems[2];
+                    continue;
+                }
+                else {
+                    trans_pkgs += "\n" + trans_elems[0];
+                    continue;
+                }
+
+            }
+
+            // Dump the last record
+            if (!trans_date.isEmpty()) {
+                // Dump previously collected entries
+                ui->recover_list_transactions->setRowCount(row+1);
+                transactions.push_back(trans_date);
+                ui->recover_list_transactions->setItem(row, 0, new QTableWidgetItem(trans_date));
+                ui->recover_list_transactions->setItem(row, 1, new QTableWidgetItem(trans_pkgs));
+                // Do not allow the user to change the line - just selection.
+                ui->recover_list_transactions->item(row, 0)->setFlags
+                    (Qt::ItemIsEnabled | Qt::ItemIsSelectable);
+                ui->recover_list_transactions->item(row, 1)->setFlags
+                    (Qt::ItemIsEnabled | Qt::ItemIsSelectable);
+                row++;
+            }
+
+            ui->urpm_status->setText(tr("Processing of data completed."));
+            procmode = LISTTRANSACTIONSHANDLED;
+            break;
 		case ORPHANSEARCH:
 			// Cleanup
 			while (orphans.size())
@@ -104,14 +222,16 @@
 			// Begin of processing the data
 			termlog("Processing ...");
 			ui->urpm_status->setText(tr("Processing data ..."));
-			for (int i = 0; proc.canReadLine(); ++i) {
-				QString neworph(proc.readLine());
+            i=0;
+            while (!in.atEnd()) {
+                QString neworph = in.readLine();
 				ui->orph_results->setRowCount(i+1);
 				orphans.push_back(neworph);
 				ui->orph_results->setItem(i, 0, new QTableWidgetItem(neworph));
 				// Do not allow the user to change the line - just selection.
 				ui->orph_results->item(i, 0)->setFlags
 					(Qt::ItemIsEnabled | Qt::ItemIsSelectable);
+                i++;
 			}
 			ui->urpm_status->setText(tr("Processing of data completed."));
 			procmode = ORPHANSEARCHHANDLED;
@@ -121,13 +241,15 @@
 			{
 				termlog("Processing ...");
 				ui->urpm_status->setText(tr("Processing data ..."));
-				for (int i = 0; proc.canReadLine(); ++i) {
-					QString newres(proc.readLine());
+                i=0;
+                while (!in.atEnd()) {
+                    QString newres = in.readLine();
 					ui->find_results->setRowCount(i+1);
 					ui->find_results->setItem(i, 0, new QTableWidgetItem(newres));
 					// Do not allow the user to change the line - just selection.
 					ui->find_results->item(i, 0)->setFlags
 						(Qt::ItemIsEnabled | Qt::ItemIsSelectable);
+                    i++;
 				}
 				ui->urpm_status->setText(tr("Processing of data completed."));
 				procmode = ORPHANSEARCHHANDLED;
@@ -140,13 +262,15 @@
 			// Begin of processing the dependencies
 			termlog("Processing dependencies...");
 			ui->urpm_status->setText(tr("Processing dependencies data ..."));
-			for (int i = 0; proc.canReadLine(); ++i) {
-				QString dep(proc.readLine());
+            i=0;
+            while (!in.atEnd()) {
+                QString dep = in.readLine();
 				if ((orphans.indexOf(dep) == -1) || (selectedOrphans.indexOf(" " + dep) > -1))
 					// either not installed or already in remove list
 					continue;
 				dependencies += " ";
 				dependencies += dep;
+                i++;
 			}
 			ui->urpm_status->setText(tr("Processing of data completed."));
 			procmode = PACKAGEDEPENDENCIESHANDLED;
@@ -161,6 +285,11 @@
 {
 	proc.close();
 	switch(procmode) {
+        case LISTTRANSACTIONS:
+            // Clean up the table - urpmq.recover --list did not return any transactions that can be reverted
+            ui->recover_list_transactions->setRowCount(0);
+            ui->urpm_status->setText(tr("No transactions to rollback found."));
+            break;
 		case ORPHANSEARCH:
 			// Clean up the table - urpmq --auto-orphans did not find anything
 			ui->orph_results->setRowCount(0);
@@ -179,7 +308,7 @@
 		default:
 			break;
 	}
-	procmode = NONE;
+    //procmode = NONE;
 }
 
 /// prepares the data for a package action - e.g. searches for dependencies.
@@ -297,6 +426,62 @@
 }
 
 /**
+ * rollback()
+ * rollbacks the package base to a given date
+ */
+void qt4urpm::rollback()
+{
+    string cmd("/usr/sbin/urpmi.recover --auto --rollback ");
+    QList<QTableWidgetItem *> packagelist(ui->recover_list_transactions->selectedItems());
+
+    if (packagelist.size() == 0) {
+        ui->urpm_status->setText(tr("No transaction selected"));
+        return;
+    }
+
+    QString targetTr = packagelist.at(0)->data(Qt::DisplayRole).toString();
+
+    // Provide user with the info - which packages will be
+    // removed and downgraded
+    // Show a dialog to ask the user whether he/she really wants to rollback
+    // to specified checkpoint
+    QString buf("<qt><p><b>");
+    buf += tr("You are about to do a system wide change!");
+    buf += "<br><br><i>";
+    buf += tr("Revert all transactions starting with the following one:");
+    buf += "</i></b></p><p>";
+    buf += targetTr;
+    buf += "</p><p><b>";
+    buf += tr("Are you sure you want to rollback?");
+    buf += "</b></p></qt>";
+
+    mydialog = new Dialog(buf);
+    mydialog->setAttribute(Qt::WA_QuitOnClose);
+    mydialog->setWindowTitle(tr("Warning!"));
+    mydialog->exec();
+
+    // Do nothing, if user canceled
+    if (mydialog->getStatus()) {
+        ui->urpm_status->setText(tr("Action canceled by user!"));
+        return;
+    }
+
+    cmd += "'" + targetTr.toStdString() + "'";
+    termlog(cmd);
+    if (system(cmd.c_str()) != 0) {
+        list_transactions();
+        termlog("An error happend during command execution.");
+        ui->urpm_status->setText(tr("WARNING: Error during execution!"));
+    } else {
+        list_transactions();
+        termlog("The system has been rolled back to " + targetTr.toStdString());
+        ui->urpm_status->setText(
+            tr("The system has been rolled back!"));
+    }
+}
+
+
+/**
  * remove()
  * uninstalls the selected package via urpme
  * \note this function may only be called via prepareRemove(), to ensure the
diff -Naubr qt4urpm-1.0.1.orig/qt4urpm.h qt4urpm-1.0.1/qt4urpm.h
--- qt4urpm-1.0.1.orig/qt4urpm.h	2014-04-01 14:19:56.272645899 +0400
+++ qt4urpm-1.0.1/qt4urpm.h	2014-04-01 13:40:25.000000000 +0400
@@ -37,6 +37,8 @@
 #include "dialog.h"
 
 enum {
+    LISTTRANSACTIONS,
+    LISTTRANSACTIONSHANDLED,
 	ORPHANSEARCH,
 	ORPHANSEARCHHANDLED,
 	PACKAGESEARCH,
@@ -60,6 +62,7 @@
 	~qt4urpm();
 
 public slots:
+    void list_transactions();
 	void orphansearch();
 	void packagesearch();
 	void processData();
@@ -72,6 +75,7 @@
 	void installAct() {install();}
 	void buttonAct();
 	void changeTooltip(int r, int c);
+    void rollback();
 
 	/// standardized terminal output
 	static void termlog(std::string const str)
@@ -93,11 +97,14 @@
 	int                procmode;
 
 	QVector<QString>   orphans;
+    QVector<QString>   transactions;
 	QString            selectedOrphans;
 	QString            dependencies;
 	QString            selectedInstall;
 	Dialog           * mydialog;
 
+    char             * procout;
+
 };
 
 #endif // QT4URPMI_H
diff -Naubr qt4urpm-1.0.1.orig/qt4urpm.pro.user qt4urpm-1.0.1/qt4urpm.pro.user
--- qt4urpm-1.0.1.orig/qt4urpm.pro.user	1970-01-01 03:00:00.000000000 +0300
+++ qt4urpm-1.0.1/qt4urpm.pro.user	2014-04-01 13:27:04.000000000 +0400
@@ -0,0 +1,260 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE QtCreatorProject>
+<!-- Written by QtCreator 3.0.1, 2014-04-01T13:27:04. -->
+<qtcreator>
+ <data>
+  <variable>ProjectExplorer.Project.ActiveTarget</variable>
+  <value type="int">0</value>
+ </data>
+ <data>
+  <variable>ProjectExplorer.Project.EditorSettings</variable>
+  <valuemap type="QVariantMap">
+   <value type="bool" key="EditorConfiguration.AutoIndent">true</value>
+   <value type="bool" key="EditorConfiguration.AutoSpacesForTabs">false</value>
+   <value type="bool" key="EditorConfiguration.CamelCaseNavigation">true</value>
+   <valuemap type="QVariantMap" key="EditorConfiguration.CodeStyle.0">
+    <value type="QString" key="language">Cpp</value>
+    <valuemap type="QVariantMap" key="value">
+     <value type="QByteArray" key="CurrentPreferences">CppGlobal</value>
+    </valuemap>
+   </valuemap>
+   <valuemap type="QVariantMap" key="EditorConfiguration.CodeStyle.1">
+    <value type="QString" key="language">QmlJS</value>
+    <valuemap type="QVariantMap" key="value">
+     <value type="QByteArray" key="CurrentPreferences">QmlJSGlobal</value>
+    </valuemap>
+   </valuemap>
+   <value type="int" key="EditorConfiguration.CodeStyle.Count">2</value>
+   <value type="QByteArray" key="EditorConfiguration.Codec">UTF-8</value>
+   <value type="bool" key="EditorConfiguration.ConstrainTooltips">false</value>
+   <value type="int" key="EditorConfiguration.IndentSize">4</value>
+   <value type="bool" key="EditorConfiguration.KeyboardTooltips">false</value>
+   <value type="bool" key="EditorConfiguration.MouseNavigation">true</value>
+   <value type="int" key="EditorConfiguration.PaddingMode">1</value>
+   <value type="bool" key="EditorConfiguration.ScrollWheelZooming">true</value>
+   <value type="int" key="EditorConfiguration.SmartBackspaceBehavior">0</value>
+   <value type="bool" key="EditorConfiguration.SpacesForTabs">true</value>
+   <value type="int" key="EditorConfiguration.TabKeyBehavior">0</value>
+   <value type="int" key="EditorConfiguration.TabSize">8</value>
+   <value type="bool" key="EditorConfiguration.UseGlobal">true</value>
+   <value type="int" key="EditorConfiguration.Utf8BomBehavior">1</value>
+   <value type="bool" key="EditorConfiguration.addFinalNewLine">true</value>
+   <value type="bool" key="EditorConfiguration.cleanIndentation">true</value>
+   <value type="bool" key="EditorConfiguration.cleanWhitespace">true</value>
+   <value type="bool" key="EditorConfiguration.inEntireDocument">false</value>
+  </valuemap>
+ </data>
+ <data>
+  <variable>ProjectExplorer.Project.PluginSettings</variable>
+  <valuemap type="QVariantMap"/>
+ </data>
+ <data>
+  <variable>ProjectExplorer.Project.Target.0</variable>
+  <valuemap type="QVariantMap">
+   <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Desktop</value>
+   <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName">Desktop</value>
+   <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">{096333a8-ed78-4104-84ae-fed2aed7d146}</value>
+   <value type="int" key="ProjectExplorer.Target.ActiveBuildConfiguration">0</value>
+   <value type="int" key="ProjectExplorer.Target.ActiveDeployConfiguration">0</value>
+   <value type="int" key="ProjectExplorer.Target.ActiveRunConfiguration">0</value>
+   <valuemap type="QVariantMap" key="ProjectExplorer.Target.BuildConfiguration.0">
+    <value type="QString" key="ProjectExplorer.BuildConfiguration.BuildDirectory">/home/tester/qt4urpm/build-qt4urpm-Desktop-Debug</value>
+    <valuemap type="QVariantMap" key="ProjectExplorer.BuildConfiguration.BuildStepList.0">
+     <valuemap type="QVariantMap" key="ProjectExplorer.BuildStepList.Step.0">
+      <value type="bool" key="ProjectExplorer.BuildStep.Enabled">true</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">qmake</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">QtProjectManager.QMakeBuildStep</value>
+      <value type="bool" key="QtProjectManager.QMakeBuildStep.LinkQmlDebuggingLibrary">false</value>
+      <value type="bool" key="QtProjectManager.QMakeBuildStep.LinkQmlDebuggingLibraryAuto">true</value>
+      <value type="QString" key="QtProjectManager.QMakeBuildStep.QMakeArguments"></value>
+      <value type="bool" key="QtProjectManager.QMakeBuildStep.QMakeForced">false</value>
+     </valuemap>
+     <valuemap type="QVariantMap" key="ProjectExplorer.BuildStepList.Step.1">
+      <value type="bool" key="ProjectExplorer.BuildStep.Enabled">true</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Сборка</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">Qt4ProjectManager.MakeStep</value>
+      <valuelist type="QVariantList" key="Qt4ProjectManager.MakeStep.AutomaticallyAddedMakeArguments">
+       <value type="QString">-w</value>
+       <value type="QString">-r</value>
+      </valuelist>
+      <value type="bool" key="Qt4ProjectManager.MakeStep.Clean">false</value>
+      <value type="QString" key="Qt4ProjectManager.MakeStep.MakeArguments"></value>
+      <value type="QString" key="Qt4ProjectManager.MakeStep.MakeCommand"></value>
+     </valuemap>
+     <value type="int" key="ProjectExplorer.BuildStepList.StepsCount">2</value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Сборка</value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">ProjectExplorer.BuildSteps.Build</value>
+    </valuemap>
+    <valuemap type="QVariantMap" key="ProjectExplorer.BuildConfiguration.BuildStepList.1">
+     <valuemap type="QVariantMap" key="ProjectExplorer.BuildStepList.Step.0">
+      <value type="bool" key="ProjectExplorer.BuildStep.Enabled">true</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Сборка</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">Qt4ProjectManager.MakeStep</value>
+      <valuelist type="QVariantList" key="Qt4ProjectManager.MakeStep.AutomaticallyAddedMakeArguments">
+       <value type="QString">-w</value>
+       <value type="QString">-r</value>
+      </valuelist>
+      <value type="bool" key="Qt4ProjectManager.MakeStep.Clean">true</value>
+      <value type="QString" key="Qt4ProjectManager.MakeStep.MakeArguments">clean</value>
+      <value type="QString" key="Qt4ProjectManager.MakeStep.MakeCommand"></value>
+     </valuemap>
+     <value type="int" key="ProjectExplorer.BuildStepList.StepsCount">1</value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Очистка</value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">ProjectExplorer.BuildSteps.Clean</value>
+    </valuemap>
+    <value type="int" key="ProjectExplorer.BuildConfiguration.BuildStepListCount">2</value>
+    <value type="bool" key="ProjectExplorer.BuildConfiguration.ClearSystemEnvironment">false</value>
+    <valuelist type="QVariantList" key="ProjectExplorer.BuildConfiguration.UserEnvironmentChanges"/>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Отладка</value>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">Qt4ProjectManager.Qt4BuildConfiguration</value>
+    <value type="int" key="Qt4ProjectManager.Qt4BuildConfiguration.BuildConfiguration">2</value>
+    <value type="bool" key="Qt4ProjectManager.Qt4BuildConfiguration.UseShadowBuild">true</value>
+   </valuemap>
+   <valuemap type="QVariantMap" key="ProjectExplorer.Target.BuildConfiguration.1">
+    <value type="QString" key="ProjectExplorer.BuildConfiguration.BuildDirectory">/home/tester/qt4urpm/build-qt4urpm-Desktop-Release</value>
+    <valuemap type="QVariantMap" key="ProjectExplorer.BuildConfiguration.BuildStepList.0">
+     <valuemap type="QVariantMap" key="ProjectExplorer.BuildStepList.Step.0">
+      <value type="bool" key="ProjectExplorer.BuildStep.Enabled">true</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">qmake</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">QtProjectManager.QMakeBuildStep</value>
+      <value type="bool" key="QtProjectManager.QMakeBuildStep.LinkQmlDebuggingLibrary">false</value>
+      <value type="bool" key="QtProjectManager.QMakeBuildStep.LinkQmlDebuggingLibraryAuto">true</value>
+      <value type="QString" key="QtProjectManager.QMakeBuildStep.QMakeArguments"></value>
+      <value type="bool" key="QtProjectManager.QMakeBuildStep.QMakeForced">false</value>
+     </valuemap>
+     <valuemap type="QVariantMap" key="ProjectExplorer.BuildStepList.Step.1">
+      <value type="bool" key="ProjectExplorer.BuildStep.Enabled">true</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Сборка</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">Qt4ProjectManager.MakeStep</value>
+      <valuelist type="QVariantList" key="Qt4ProjectManager.MakeStep.AutomaticallyAddedMakeArguments">
+       <value type="QString">-w</value>
+       <value type="QString">-r</value>
+      </valuelist>
+      <value type="bool" key="Qt4ProjectManager.MakeStep.Clean">false</value>
+      <value type="QString" key="Qt4ProjectManager.MakeStep.MakeArguments"></value>
+      <value type="QString" key="Qt4ProjectManager.MakeStep.MakeCommand"></value>
+     </valuemap>
+     <value type="int" key="ProjectExplorer.BuildStepList.StepsCount">2</value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Сборка</value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">ProjectExplorer.BuildSteps.Build</value>
+    </valuemap>
+    <valuemap type="QVariantMap" key="ProjectExplorer.BuildConfiguration.BuildStepList.1">
+     <valuemap type="QVariantMap" key="ProjectExplorer.BuildStepList.Step.0">
+      <value type="bool" key="ProjectExplorer.BuildStep.Enabled">true</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Сборка</value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+      <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">Qt4ProjectManager.MakeStep</value>
+      <valuelist type="QVariantList" key="Qt4ProjectManager.MakeStep.AutomaticallyAddedMakeArguments">
+       <value type="QString">-w</value>
+       <value type="QString">-r</value>
+      </valuelist>
+      <value type="bool" key="Qt4ProjectManager.MakeStep.Clean">true</value>
+      <value type="QString" key="Qt4ProjectManager.MakeStep.MakeArguments">clean</value>
+      <value type="QString" key="Qt4ProjectManager.MakeStep.MakeCommand"></value>
+     </valuemap>
+     <value type="int" key="ProjectExplorer.BuildStepList.StepsCount">1</value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Очистка</value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">ProjectExplorer.BuildSteps.Clean</value>
+    </valuemap>
+    <value type="int" key="ProjectExplorer.BuildConfiguration.BuildStepListCount">2</value>
+    <value type="bool" key="ProjectExplorer.BuildConfiguration.ClearSystemEnvironment">false</value>
+    <valuelist type="QVariantList" key="ProjectExplorer.BuildConfiguration.UserEnvironmentChanges"/>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Выпуск</value>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">Qt4ProjectManager.Qt4BuildConfiguration</value>
+    <value type="int" key="Qt4ProjectManager.Qt4BuildConfiguration.BuildConfiguration">0</value>
+    <value type="bool" key="Qt4ProjectManager.Qt4BuildConfiguration.UseShadowBuild">true</value>
+   </valuemap>
+   <value type="int" key="ProjectExplorer.Target.BuildConfigurationCount">2</value>
+   <valuemap type="QVariantMap" key="ProjectExplorer.Target.DeployConfiguration.0">
+    <valuemap type="QVariantMap" key="ProjectExplorer.BuildConfiguration.BuildStepList.0">
+     <value type="int" key="ProjectExplorer.BuildStepList.StepsCount">0</value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Установка</value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+     <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">ProjectExplorer.BuildSteps.Deploy</value>
+    </valuemap>
+    <value type="int" key="ProjectExplorer.BuildConfiguration.BuildStepListCount">1</value>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">Локальная установка</value>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">ProjectExplorer.DefaultDeployConfiguration</value>
+   </valuemap>
+   <value type="int" key="ProjectExplorer.Target.DeployConfigurationCount">1</value>
+   <valuemap type="QVariantMap" key="ProjectExplorer.Target.PluginSettings"/>
+   <valuemap type="QVariantMap" key="ProjectExplorer.Target.RunConfiguration.0">
+    <valuelist type="QVariantList" key="Analyzer.Valgrind.AddedSuppressionFiles"/>
+    <value type="bool" key="Analyzer.Valgrind.Callgrind.CollectBusEvents">false</value>
+    <value type="bool" key="Analyzer.Valgrind.Callgrind.CollectSystime">false</value>
+    <value type="bool" key="Analyzer.Valgrind.Callgrind.EnableBranchSim">false</value>
+    <value type="bool" key="Analyzer.Valgrind.Callgrind.EnableCacheSim">false</value>
+    <value type="bool" key="Analyzer.Valgrind.Callgrind.EnableEventToolTips">true</value>
+    <value type="double" key="Analyzer.Valgrind.Callgrind.MinimumCostRatio">0.01</value>
+    <value type="double" key="Analyzer.Valgrind.Callgrind.VisualisationMinimumCostRatio">10</value>
+    <value type="bool" key="Analyzer.Valgrind.FilterExternalIssues">true</value>
+    <value type="int" key="Analyzer.Valgrind.LeakCheckOnFinish">1</value>
+    <value type="int" key="Analyzer.Valgrind.NumCallers">25</value>
+    <valuelist type="QVariantList" key="Analyzer.Valgrind.RemovedSuppressionFiles"/>
+    <value type="int" key="Analyzer.Valgrind.SelfModifyingCodeDetection">1</value>
+    <value type="bool" key="Analyzer.Valgrind.Settings.UseGlobalSettings">true</value>
+    <value type="bool" key="Analyzer.Valgrind.ShowReachable">false</value>
+    <value type="bool" key="Analyzer.Valgrind.TrackOrigins">true</value>
+    <value type="QString" key="Analyzer.Valgrind.ValgrindExecutable">valgrind</value>
+    <valuelist type="QVariantList" key="Analyzer.Valgrind.VisibleErrorKinds">
+     <value type="int">0</value>
+     <value type="int">1</value>
+     <value type="int">2</value>
+     <value type="int">3</value>
+     <value type="int">4</value>
+     <value type="int">5</value>
+     <value type="int">6</value>
+     <value type="int">7</value>
+     <value type="int">8</value>
+     <value type="int">9</value>
+     <value type="int">10</value>
+     <value type="int">11</value>
+     <value type="int">12</value>
+     <value type="int">13</value>
+     <value type="int">14</value>
+    </valuelist>
+    <value type="int" key="PE.EnvironmentAspect.Base">2</value>
+    <valuelist type="QVariantList" key="PE.EnvironmentAspect.Changes"/>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.DefaultDisplayName">qt4urpm</value>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.DisplayName"></value>
+    <value type="QString" key="ProjectExplorer.ProjectConfiguration.Id">Qt4ProjectManager.Qt4RunConfiguration:/home/tester/qt4urpm/qt4urpm-1.0.1/qt4urpm.pro</value>
+    <value type="QString" key="Qt4ProjectManager.Qt4RunConfiguration.CommandLineArguments"></value>
+    <value type="QString" key="Qt4ProjectManager.Qt4RunConfiguration.ProFile">qt4urpm.pro</value>
+    <value type="bool" key="Qt4ProjectManager.Qt4RunConfiguration.UseDyldImageSuffix">false</value>
+    <value type="bool" key="Qt4ProjectManager.Qt4RunConfiguration.UseTerminal">false</value>
+    <value type="QString" key="Qt4ProjectManager.Qt4RunConfiguration.UserWorkingDirectory"></value>
+    <value type="uint" key="RunConfiguration.QmlDebugServerPort">3768</value>
+    <value type="bool" key="RunConfiguration.UseCppDebugger">true</value>
+    <value type="bool" key="RunConfiguration.UseCppDebuggerAuto">false</value>
+    <value type="bool" key="RunConfiguration.UseMultiProcess">false</value>
+    <value type="bool" key="RunConfiguration.UseQmlDebugger">false</value>
+    <value type="bool" key="RunConfiguration.UseQmlDebuggerAuto">true</value>
+   </valuemap>
+   <value type="int" key="ProjectExplorer.Target.RunConfigurationCount">1</value>
+  </valuemap>
+ </data>
+ <data>
+  <variable>ProjectExplorer.Project.TargetCount</variable>
+  <value type="int">1</value>
+ </data>
+ <data>
+  <variable>ProjectExplorer.Project.Updater.EnvironmentId</variable>
+  <value type="QByteArray">{a80be3c3-f889-4df9-aeb0-c54f34d607cf}</value>
+ </data>
+ <data>
+  <variable>ProjectExplorer.Project.Updater.FileVersion</variable>
+  <value type="int">15</value>
+ </data>
+</qtcreator>
diff -Naubr qt4urpm-1.0.1.orig/qt4urpm.ui qt4urpm-1.0.1/qt4urpm.ui
--- qt4urpm-1.0.1.orig/qt4urpm.ui	2014-04-01 14:19:56.272645899 +0400
+++ qt4urpm-1.0.1/qt4urpm.ui	2014-04-01 14:21:31.581268212 +0400
@@ -109,7 +109,7 @@
        <enum>QTabWidget::Rounded</enum>
       </property>
       <property name="currentIndex">
-       <number>0</number>
+       <number>1</number>
       </property>
       <widget class="QWidget" name="tab_orphaned">
        <property name="sizePolicy">
@@ -144,7 +144,16 @@
         <property name="sizeConstraint">
          <enum>QLayout::SetMinAndMaxSize</enum>
         </property>
-        <property name="margin">
+        <property name="leftMargin">
+         <number>0</number>
+        </property>
+        <property name="topMargin">
+         <number>0</number>
+        </property>
+        <property name="rightMargin">
+         <number>0</number>
+        </property>
+        <property name="bottomMargin">
          <number>0</number>
         </property>
         <item>
@@ -289,6 +298,197 @@
         </item>
        </layout>
       </widget>
+      <widget class="QWidget" name="tab_rollback">
+       <property name="sizePolicy">
+        <sizepolicy hsizetype="Expanding" vsizetype="Expanding">
+         <horstretch>1</horstretch>
+         <verstretch>1</verstretch>
+        </sizepolicy>
+       </property>
+       <property name="minimumSize">
+        <size>
+         <width>555</width>
+         <height>320</height>
+        </size>
+       </property>
+       <property name="sizeIncrement">
+        <size>
+         <width>1</width>
+         <height>1</height>
+        </size>
+       </property>
+       <attribute name="icon">
+        <iconset>
+         <normaloff>/usr/share/icons/packaging_section.png</normaloff>/usr/share/icons/packaging_section.png</iconset>
+       </attribute>
+       <attribute name="title">
+        <string>urpmi recover</string>
+       </attribute>
+       <layout class="QVBoxLayout" name="verticalLayout_31">
+        <property name="spacing">
+         <number>3</number>
+        </property>
+        <property name="sizeConstraint">
+         <enum>QLayout::SetMinAndMaxSize</enum>
+        </property>
+        <property name="leftMargin">
+         <number>0</number>
+        </property>
+        <property name="topMargin">
+         <number>0</number>
+        </property>
+        <property name="rightMargin">
+         <number>0</number>
+        </property>
+        <property name="bottomMargin">
+         <number>0</number>
+        </property>
+        <item>
+         <widget class="QTableWidget" name="recover_list_transactions">
+          <property name="sizePolicy">
+           <sizepolicy hsizetype="Expanding" vsizetype="Expanding">
+            <horstretch>1</horstretch>
+            <verstretch>1</verstretch>
+           </sizepolicy>
+          </property>
+          <property name="minimumSize">
+           <size>
+            <width>550</width>
+            <height>255</height>
+           </size>
+          </property>
+          <property name="sizeIncrement">
+           <size>
+            <width>1</width>
+            <height>1</height>
+           </size>
+          </property>
+          <property name="baseSize">
+           <size>
+            <width>0</width>
+            <height>0</height>
+           </size>
+          </property>
+          <property name="mouseTracking">
+           <bool>true</bool>
+          </property>
+          <property name="locale">
+           <locale language="English" country="UnitedStates"/>
+          </property>
+          <property name="selectionMode">
+           <enum>QAbstractItemView::SingleSelection</enum>
+          </property>
+          <property name="selectionBehavior">
+           <enum>QAbstractItemView::SelectRows</enum>
+          </property>
+          <attribute name="horizontalHeaderCascadingSectionResizes">
+           <bool>true</bool>
+          </attribute>
+          <attribute name="horizontalHeaderDefaultSectionSize">
+           <number>300</number>
+          </attribute>
+          <attribute name="horizontalHeaderHighlightSections">
+           <bool>true</bool>
+          </attribute>
+          <attribute name="horizontalHeaderMinimumSectionSize">
+           <number>300</number>
+          </attribute>
+          <attribute name="verticalHeaderCascadingSectionResizes">
+           <bool>true</bool>
+          </attribute>
+          <attribute name="verticalHeaderDefaultSectionSize">
+           <number>60</number>
+          </attribute>
+          <column>
+           <property name="text">
+            <string>Date</string>
+           </property>
+          </column>
+          <column>
+           <property name="text">
+            <string>Packages</string>
+           </property>
+          </column>
+         </widget>
+        </item>
+        <item>
+         <layout class="QHBoxLayout" name="trs_horizontalLayout">
+          <property name="spacing">
+           <number>3</number>
+          </property>
+          <property name="sizeConstraint">
+           <enum>QLayout::SetMinAndMaxSize</enum>
+          </property>
+          <property name="leftMargin">
+           <number>2</number>
+          </property>
+          <property name="rightMargin">
+           <number>2</number>
+          </property>
+          <item>
+           <widget class="QPushButton" name="trs_update">
+            <property name="sizePolicy">
+             <sizepolicy hsizetype="Fixed" vsizetype="Fixed">
+              <horstretch>0</horstretch>
+              <verstretch>0</verstretch>
+             </sizepolicy>
+            </property>
+            <property name="minimumSize">
+             <size>
+              <width>0</width>
+              <height>20</height>
+             </size>
+            </property>
+            <property name="locale">
+             <locale language="English" country="UnitedStates"/>
+            </property>
+            <property name="text">
+             <string>Update List</string>
+            </property>
+           </widget>
+          </item>
+          <item>
+           <spacer name="trs_horizontalSpacer">
+            <property name="orientation">
+             <enum>Qt::Horizontal</enum>
+            </property>
+            <property name="sizeHint" stdset="0">
+             <size>
+              <width>40</width>
+              <height>20</height>
+             </size>
+            </property>
+           </spacer>
+          </item>
+          <item>
+           <widget class="QPushButton" name="trs_rollback">
+            <property name="sizePolicy">
+             <sizepolicy hsizetype="Fixed" vsizetype="Fixed">
+              <horstretch>0</horstretch>
+              <verstretch>0</verstretch>
+             </sizepolicy>
+            </property>
+            <property name="minimumSize">
+             <size>
+              <width>0</width>
+              <height>20</height>
+             </size>
+            </property>
+            <property name="toolTip">
+             <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Rollback all transactions starting from the selected one.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
+            </property>
+            <property name="locale">
+             <locale language="English" country="UnitedStates"/>
+            </property>
+            <property name="text">
+             <string>Rollback</string>
+            </property>
+           </widget>
+          </item>
+         </layout>
+        </item>
+       </layout>
+      </widget>
       <widget class="QWidget" name="tab_find">
        <property name="sizePolicy">
         <sizepolicy hsizetype="Expanding" vsizetype="Expanding">
@@ -469,7 +669,7 @@
      <x>0</x>
      <y>0</y>
      <width>590</width>
-     <height>26</height>
+     <height>19</height>
     </rect>
    </property>
    <property name="locale">
